### PR TITLE
Allow to offload certificate validation when using BoringSSL

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

BoringSSL supports offloading certificate validation to a different thread. This is useful as it may need to do blocking operations and so may block the EventLoop.

Modification:

- Adjust ReferenceCountedOpenSslEngine to correctly handle offloaded certificate validation (just as we already have code for certificate selection).

Result:

Be able to offload certificate validation when using BoringSSL.

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 1]
```diff
@@ -737,7 +737,7 @@ public final SSLEngineResult wrap(
                     // Flush any data that may have been written implicitly during the handshake by OpenSSL.
                     bytesProduced = SSL.bioFlushByteBuffer(networkBIO);
 
-                    if (bytesProduced > 0 && handshakeException != null) {
```
Comment (johnou) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: Why was this removed? I recall some special checks for Android's "broken" SSL implementation..

Comment (johnou) [reply handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: nvm looks like another code path https://github.com/netty/netty/pull/7761

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 2]
```diff
@@ -746,7 +746,9 @@ public final SSLEngineResult wrap(
                         // We produced / consumed some data during the handshake, signal back to the caller.
                         // If there is a handshake exception and we have produced data, we should send the data before
                         // we allow handshake() to throw the handshake exception.
-                        return newResult(NEED_WRAP, 0, bytesProduced);
+                        //
+                        // unwrap(...) will then ensure we propagate the handshake error back to the user.
+                        return newResult(NEED_UNWRAP, 0, bytesProduced);
```
Comment (johnou) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: cc @Scottmitch could I trouble you to check this out since it has your name on it :wink:

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 3]
```diff
@@ -889,7 +891,8 @@ public final SSLEngineResult wrap(
                             // to write encrypted data to. This is an OVERFLOW condition.
                             // [1] https://www.openssl.org/docs/manmaster/ssl/SSL_write.html
                             return newResult(BUFFER_OVERFLOW, status, bytesConsumed, bytesProduced);
-                        } else if (sslError == SSL.SSL_ERROR_WANT_X509_LOOKUP) {
+                        } else if (sslError == SSL.SSL_ERROR_WANT_X509_LOOKUP ||
```
Comment (carl-mastrangelo) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: nit: the binary op `||` is wrapped on the first line here, but on the second a few lines down.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 2 above)
Comment (johnou) [reply handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @normanmaurer can we add a test for this?

Comment (normanmaurer) [reply handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: its covered by the tests already ;)

**Comment by normanmaurer** [comment]
Let me revert this as there seems to be some test errors on my local machine :/ 

Reverted as 33e2f5609d

**Comment by normanmaurer** [comment]
Re-opened as https://github.com/netty/netty/pull/8974
